### PR TITLE
fix: @provides validation for object types

### PR DESCRIPTION
**Pull request by lennyburdette**
_Please feel free to close this if I've misunderstood the federation spec!_

fix: @provides validation for object types
  
The previous post-composition validators `providesFieldsMissingExternal` and
`providesFieldsSelectInvalidType` conflicted with the [federation spec][spec].

In the case of "missing external" validations, the previous code looked for
the provided field on the wrong type. Using the following example:

```gql
type A {
  b: C @provides(fields: "d")
}

type C {
  d: Int
}
```

The validator was looking for field `d` on type `A`. It now looks at the type
of `b` to correctly lookup field `d` on type `C`.

Similarly, the "fields select invalid type" validator looked up the provided
field incorrectly. The fix is basically the same in both validators.

[spec]:https://www.apollographql.com/docs/apollo-server/federation/advanced-features/#using-denormalized-data

### CODE REVIEW

File: packages/apollo-federation/src/composition/__tests__/compose.test.ts [hunk 1]
```diff
@@ -994,9 +994,12 @@ describe('composeServices', () => {
         expect(errors).toHaveLength(0);
 
         const review = schema.getType('Review') as GraphQLObjectType;
-        expect(
-          review.getFields()['product'].federation.provides,
-        ).toMatchInlineSnapshot(`sku`);
+        expect(review.getFields()['product'].federation).toMatchInlineSnapshot(`
+          Object {
+            "provides": sku,
```
Comment (JacksonKearl) [inline packages/apollo-federation/src/composition/__tests__/compose.test.ts]: Is there a reason `sku` is not in quotes? I'm not very familiar with the snapshot strategy.

**Comment by JacksonKearl** [comment]
Thanks for this! This is absolutely something that needed to be fixed.

File: packages/apollo-federation/src/composition/__tests__/compose.test.ts (hunk 1 above)
Comment (JacksonKearl) [reply packages/apollo-federation/src/composition/__tests__/compose.test.ts]: Ah I see. It's the snapshot serializer's description of a more complicated GraphQL field object.

**Comment by JakeDawkins** [comment]
Thanks!! I'll take the blame on this one not working originally. I had a misunderstanding of the spec when I wrote it 😅

Either way, great work on this! Approved ✅💯

**Pull request merged**